File: ios/LaneShadow/Views/Atoms/ThemeToggle.swift
import LaneShadowTheme
import SwiftUI

struct ThemeToggle: View {
    @Environment(\.theme) private var theme

    @Binding var isPressed: Bool
    let label: String
    let variant: ThemeToggleVariant
    let size: ThemeToggleSize
    let iconName: String?
    let isEnabled: Bool

    init(
        _ label: String,
        isPressed: Binding<Bool>,
        variant: ThemeToggleVariant = .default,
        size: ThemeToggleSize = .md,
        iconName: String? = nil,
        isEnabled: Bool = true
    ) {
        self.label = label
        _isPressed = isPressed
        self.variant = variant
        self.size = size
        self.iconName = iconName
        self.isEnabled = isEnabled
    }

    var body: some View {
        Button {
            guard isEnabled else { return }
            isPressed.toggle()
        } label: {
            HStack(spacing: theme.space.sm) {
                if let iconName {
                    ThemeIcon(
                        name: iconName,
                        size: theme.space.md,
                        color: isPressed ? theme.colors.onSurface.default : theme.colors.muted.default
                    )
                }

                ThemeText(
                    label,
                    variant: .labelMd,
                    color: isPressed ? theme.colors.onSurface.default : theme.colors.muted.default
                )
            }
            .padding(.horizontal, theme.space.md)
            .padding(.vertical, size.verticalPadding(in: theme))
            .background(backgroundColor)
            .overlay {
                RoundedRectangle(cornerRadius: theme.radius.md, style: .continuous)
                    .stroke(borderColor, lineWidth: borderColor == .clear ? 0 : 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: theme.radius.md, style: .continuous))
        }
        .buttonStyle(.plain)
        .opacity(isEnabled ? 1 : 0.7)
        .animation(.easeInOut(duration: 0.16), value: isPressed)
        .accessibilityLabel(label)
    }

    private var backgroundColor: Color {
        guard isEnabled else { return theme.colors.muted.disabled ?? theme.colors.muted.default }
        return isPressed ? theme.colors.accent.default : theme.colors.background.default
    }

    private var borderColor: Color {
        variant == .outline ? theme.colors.border.default : .clear
    }
}

enum ThemeToggleVariant: String, CaseIterable {
    case `default`
    case outline
}

enum ThemeToggleSize: String, CaseIterable {
    case sm
    case md
    case lg

    fileprivate func verticalPadding(in theme: Theme) -> CGFloat {
        switch self {
        case .sm, .md: theme.space.sm
        case .lg: theme.space.md
        }
    }
}

File: android/app/src/main/java/com/laneshadow/ui/atoms/ThemeToggle.kt
package com.laneshadow.ui.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

@Composable
fun ThemeToggle(
    pressed: Boolean,
    onPressedChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    variant: ThemeToggleVariant = ThemeToggleVariant.Default,
    size: ThemeToggleSize = ThemeToggleSize.Md,
    enabled: Boolean = true,
    iconName: String? = null,
    label: String,
) {
    val theme = LocalLaneShadowTheme.current
    val border = if (variant == ThemeToggleVariant.Outline) BorderStroke(1.dp, theme.colors.border.default) else null

    Surface(
        modifier =
            modifier.clickable(enabled = enabled) {
                onPressedChange(!pressed)
            },
        shape = RoundedCornerShape(theme.radius.md),
        color =
            when {
                !enabled -> theme.colors.muted.disabled ?: theme.colors.muted.default
                pressed -> theme.colors.accent.default
                else -> theme.colors.background.default
            },
        border = border,
        tonalElevation = if (pressed) 1.dp else 0.dp,
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(horizontal = theme.space.md, vertical = toggleVerticalPadding(theme, size)),
        ) {
            if (iconName != null) {
                IconSymbol(
                    name = iconName,
                    color = if (pressed) theme.colors.onSurface.default else theme.colors.muted.default,
                    size = 18.dp,
                )
            }
            ThemedText(
                text = label,
                variant = ThemedTextVariant.LabelMd,
                color = if (pressed) theme.colors.onSurface.default else theme.colors.muted.default,
            )
        }
    }
}

enum class ThemeToggleVariant {
    Default,
    Outline,
}

enum class ThemeToggleSize {
    Sm,
    Md,
    Lg,
}

@Composable
private fun toggleVerticalPadding(
    theme: com.laneshadow.theme.LaneShadowThemeValues,
    size: ThemeToggleSize,
) = when (size) {
    ThemeToggleSize.Sm -> theme.space.sm
    ThemeToggleSize.Lg -> theme.space.md
    ThemeToggleSize.Md -> theme.space.sm
}
